### PR TITLE
fixes #544; Added empty.xml ruleset and test cases for the same

### LINKED ISSUE #544: Qulice does not enforce any of "Empty Code" PMD rules
There are lots of rules in PMD ['empty' ruleset](https://pmd.github.io/pmd-5.4.1/pmd-java/rules/java/empty.html), but they are not mentioned in Qulice's PMD configuration.

I would excpect all of them being enforced by Qulice.

### PR DESCRIPTION
**Pull request by prahladyeri**
I've added the class `PMDEmptyTest` along with a bunch of resources to verify that each one of these `empty.xml` ruleset is validated. This is a fix for issue #544.

### CODE REVIEW

**Comment by prahladyeri** [comment]
@krzyk The build on travis is failing for the `PMDEmptyTest.emptyInitializerTest` test as below, whereas the build ran perfectly on my local machine. I don't understand this.

    PMDEmptyTest.emptyInitializerTest:86->validatePMD:217 
    Expected: a string containing "Empty initializer was found"

**Comment by krzyk** [comment]
@prahladyeri maybe you should add `writer.flush()` as was done in [PMDValidatorTest](https://github.com/teamed/qulice/blob/f99a57c13a2949f95af9cecbbb896dcccd7e7221/qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java#L303)

**Comment by prahladyeri** [comment]
@krzyk Thanks, `writer.flush()` was indeed the issue, along with the fact that I had mistakenly added the `appender` twice.

**Comment by prahladyeri** [comment]
@krzyk I have a basic technical query. When I build qulice-pmd project using `mvn clean install -Pqulice`, a lot of junk files in this pattern are created in the parent folder. Is there a way to avoid that, or is this just normal?

    qulice-pmd:findbug_excludes_5481773978200.xml

**Comment by davvd** [comment]
@prahladyeri Many thanks for the PR, let me find a reviewer for it

**Comment by davvd** [comment]
@carlosmiranda review this one,please

**Comment by krzyk** [comment]
@prahladyeri junk files were appearing because we had a bug in our code some time ago, after recent upgrade to jcabi-parent 0.40 they should disappear

**Comment by prahladyeri** [comment]
@krzyk thanks

**Comment by krzyk** [comment]
@prahladyeri If you are waiting long for review, you can inform reviewer that you are waiting. You are responsible for completing the task in 7-10 days, so it is in you interest to be proactive :)

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+import org.hamcrest.Matchers;
+import org.junit.Test;
+
+/**
+ * Test case for {@link PMDValidator} class.
+ * @author Prahlad Yeri (prahladyeri@yahoo.com)
+ * @version $Id$
+ */
+@SuppressWarnings("PMD.TooManyMethods")
+public final class PMDEmptyTest {
```
Comment (carlosmiranda) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java]: @prahladyeri Let's put all of these tests in `PMDValidatorTest`.

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    /**
+     * Makes sure that empty static initializers fail with an error.
+     * @throws Exception when something goes wrong
+     */
+    @Test
+    public void emptyStaticInitializerTest() throws Exception {
```
Comment (carlosmiranda) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java]: @prahladyeri Let's call this method `failsForEmptyStaticInitializer`, see [this page](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names) for more details.

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            )
+        );
+    }
+
+    /**
+     * Makes sure that empty statement blocks fail with an error.
+     * @throws Exception when something goes wrong
+     */
+    @Test
+    public void emptyStatementBlockTest() throws Exception {
```
Comment (carlosmiranda) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java]: @prahladyeri Let's call this method `failsForEmptyStatementBlock`, see [this page](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names) for more details.

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+            )
+        );
+    }
+
+    /**
+     * Makes sure that empty initializers fail with an error.
+     * @throws Exception when something goes wrong
+     */
+    @Test
+    public void emptyInitializerTest() throws Exception {
```
Comment (carlosmiranda) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java]: @prahladyeri Let's call this method `failsForEmptyInitializer`, see [this page](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names) for more details.

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            )
+        );
+    }
+
+    /**
+     * Makes sure that empty statement not in a loop fail with an error.
+     * @throws Exception when something goes wrong
+     */
+    @Test
+    public void emptyStatementNotInLoopTest() throws Exception {
```
Comment (carlosmiranda) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java]: @prahladyeri Let's call this method `failsForEmptyNonLoopStatement`, see [this page](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names) for more details.

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            )
+        );
+    }
+
+    /**
+     * Makes sure that empty synchronized statements fail with an error.
+     * @throws Exception when something goes wrong
+     */
+    @Test
+    public void emptySynchronizedBlockTest() throws Exception {
```
Comment (carlosmiranda) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java]: @prahladyeri Let's call this method `failsForEmptySynchronizedBlock`, see [this page](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names) for more details.

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            )
+        );
+    }
+
+    /**
+     * Makes sure that empty switch statements fail with an error.
+     * @throws Exception when something goes wrong
+     */
+    @Test
+    public void emptySwitchStmtTest() throws Exception {
```
Comment (carlosmiranda) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java]: @prahladyeri Let's call this method `failsForEmptySwitchStatement`, see [this page](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names) for more details.

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+            )
+        );
+    }
+
+    /**
+     * Makes sure that empty finally blocks fail with an error.
+     * @throws Exception when something goes wrong
+     */
+    @Test
+    public void emptyFinallyBlockTest() throws Exception {
```
Comment (carlosmiranda) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java]: @prahladyeri Let's call this method `failsForEmptyFinallyBlock`, see [this page](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names) for more details.

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+            Matchers.containsString("Avoid empty while statements")
+        );
+    }
+
+    /**
+     * Makes sure that empty if blocks fail with an error.
+     * @throws Exception when something goes wrong
+     */
+    @Test
+    public void emptyIfStmtTest() throws Exception {
```
Comment (carlosmiranda) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java]: @prahladyeri Let's call this method `failsForEmptyIfStatement`, see [this page](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names) for more details.

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+            Matchers.containsString("Avoid empty if statements")
+        );
+    }
+
+    /**
+     * Makes sure that empty catch blocks fail with an error.
+     * @throws Exception when something goes wrong
+     */
+    @Test
+    public void emptyCatchBlockTest() throws Exception {
```
Comment (carlosmiranda) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java]: @prahladyeri Let's call this method `failsForEmptyCatchBlock`, see [this page](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names) for more details.

**Comment by carlosmiranda** [comment]
@krzyk @prahladyeri sorry for the delay, holiday hangovers. :)

**Comment by carlosmiranda** [comment]
@prahladyeri Very good, I have some comments mostly related to test method naming.

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+            Matchers.containsString("Avoid empty finally blocks")
+        );
+    }
+
+    /**
+     * Makes sure that empty while statements fail with an error.
+     * @throws Exception when something goes wrong
+     */
+    @Test
+    public void emptyWhileStmtTest() throws Exception {
```
Comment (carlosmiranda) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java]: @prahladyeri Let's call this method `failsForEmptyWhileStatement
`, see [this page](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names) for more details.

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+        );
+    }
+
+    /**
+     * Validates that PMD reported given violation.
+     * @param file String containing file name to emulate.
+     * @param matcher Matching string needed in the output.
+     * @throws Exception If something wrong happens inside.
+     */
+    private void validatePMD(final String file,
```
Comment (carlosmiranda) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDEmptyTest.java]: @prahladyeri This code is duplicated from `PMDValidatorTest`, let's avoid that (DRY principle). Better to place these tests in that file.

**Comment by prahladyeri** [comment]
@krzyk The tests are failing randomly this time. I think it is because there are too many methods and the `junit` runs them asynchronously all at once. So expected string patterns (like `"Avoid empty while statements"`) are sometimes not found. Shall I try to make them run in sync?

**Comment by krzyk** [comment]
@prahladyeri this is similar problem as in checkstyle, so we will fix it in https://github.com/teamed/qulice/issues/547, also the flush should be before asserts. For now add `@NotThreadSafe` (from jcip) annotation to the test class.

**Comment by prahladyeri** [comment]
@krzyk Are you sure `@NotThreadSafe` will work? Firstly, I don't see any used examples in qulice.  According to [this](http://www.coderanch.com/t/233975/threads/java/jcip-annotation) link:

> net.jcip.annotations are intended to be used as documentation and annotations themselves doesn't enforce any thread safety or concurrency control.

So, does the `jUnit` honor `@NotThreadSafe` attribute?

**Comment by krzyk** [comment]
@prahladyeri it is a feature of maven surefire, see https://maven.apache.org/surefire/maven-surefire-plugin/examples/fork-options-and-parallel-execution.html#Parallel_Test_Execution_and_Single_Thread_Execution

**Comment by krzyk** [comment]
@prahladyeri but first try to move the tests to PMDValidatorTest class (without adding the annotation), maybe it will work, if the flush is at the appropriate place

**Comment by prahladyeri** [comment]
@krzyk Already tried moving all tests to `PMDValidatorTest` class, that didn't work. In fact, that scenario was worse as the build failed each time, with a separate class, there is at least a 50% chance of success!

**Comment by krzyk** [comment]
@prahladyeri strange, build didn't fail before, so I assume the reason is in this PR

**Comment by prahladyeri** [comment]
> @prahladyeri strange, build didn't fail before

As I said, the build has a 50% chance of success, maybe that's why it didn't fail before. All that has changed in this PR is the method names.

**Comment by prahladyeri** [comment]
@krzyk I couldn't find a good working example of `@NotThreadSafe`, so I ended up by changing the `surefire` configuration to single threading as per below configuration, which basically does the same thing (so far ten builds succeeded without issue):

     <forkCount>1</forkCount>
    <threadCount>1</threadCount>
    <parallel>classes</parallel>

Source: [StackOverflow](http://stackoverflow.com/questions/32987998/junit-and-surefire-parallel-tests-forkcount-threadcount/33047617)

@carlosmiranda Made changes as per all mentioned comments with the single exception of merging with `PMDValidator` class. I've kept the class separate as it groups all `Empty` validations into one.

**Comment by carlosmiranda** [comment]
@rultor merge please

**Comment by krzyk** [comment]
@prahladyeri I still think you should try `@NotThreadSafe` annotation, you just annotate the main test class and your done, this way you are not changing the way all other tests are working.

Here's an example:
```
import net.jcip.annotations.NotThreadSafe;

@NotThreadSafe
public final class PMDValidatorTest {
```

and in the pom:
```
        <dependency>
            <groupId>net.jcip</groupId>
            <artifactId>jcip-annotations</artifactId>
            <version>1.0</version>
            <scope>test</scope>
        </dependency>
```

**Comment by prahladyeri** [comment]
@krzyk I get the below error despite adding the dependency in `pom.xml`:

    package net.jcip.annotations does not exist

On the other hand, are you sure that the `surefire plugin` will honor the `@ThreadSafe` attribute? I don't think so, because the problem we are facing is because of multi-threading at the `METHOD` level, not the `CLASS` level. As I explained in #547 yesterday, there are some tests in `PMDValidator` class which pass only on not finding a particular string in the log output (similar to @HDouss 's case). And `surefire` needs to be told to execute these tests one by one at the method level, not all at once.

**Comment by krzyk** [comment]
@prahladyeri Yes I am sure  surefire will honor it, I've used it in other projects, and it is mentioned in surefire docs after all. The only thing that needs to be tested is if it will run each methods separately.

**Comment by HDouss** [comment]
@prahladyeri Please note that the case I described is not the case where a 
> class which pass only on not finding a particular string in the log output

But a class which expects a particular string in the log output. This is about false positives. A positive means a check violation, so an expected string in the output.

**Comment by prahladyeri** [comment]
@krzyk Okay, but can you at least explain what is the inferiority in my solution?

**Comment by prahladyeri** [comment]
@HDouss I agree, but the root cause is same in both cases, which is multi-threading:

i. Finding a string that is not expected in output.
ii. Not finding a string that is expected in output.

If you think about it, the only difference is the criteria of a acceptance/rejection, but the underlying string matching logic remains same in both cases.

**Comment by HDouss** [comment]
@prahladyeri Yes the root cause may be multithreading but the two issues are different.
In one case (#547) , you have more output than expected. The build don't fail at all. You have exactly what expected (the problem is that you shouldn't). What happened to have such issue is clear and can somehow be reproduced.
In the other case you are expecting a string that you don't find, and the build fails. Such a multithreaded scenario must produce more output than expected. So, what happened to have such issue is still mysterious until now and there is no clear explanation to it. It may (and very likely to) be related to multithreading, but we don't have any explanation. That's why I think that the two issues are different.

**Comment by krzyk** [comment]
@prahladyeri you are changing the way all current (and future) test run in this module. Changing only single class is preferred if it fixes the problem.

**Comment by krzyk** [comment]
@prahladyeri you can also add a todo puzzle in the pom.xml or in your test class, in which you can write that the tests fail and you temporarily disabled parallel tests to make them work.
(see http://www.yegor256.com/2009/03/04/pdd.html and http://www.yegor256.com/2014/04/12/puzzle-driven-development-by-roles.html)

I would write something like in the javadoc of `PMDEmptyTest`:
```
@todo #544:30min Tests below pass only when run sequentially, when they are run in parallel some of them start to fail. Please fix the tests below and remove override of maven-surefire-plugin configuration in qulice-pmd pom.xml
```

Puzzle text should contain number of minutes you estimate it can take (usually we use `30min` if we are not sure, or `15min` if we are really sure it is very small change) and description should have >= 15 words.

This way you can split a bigger task (one that doesn't fit into e.g. 30mins) into smaller ones, it is important that each PR adds some value - e.g. it can cover only happy path and add todo (or todos) for corner cases.

@HDouss @mkordas @jrdalpra @aschworer I'm not sure if you were introduced to todo puzzles and PDD before, please read this text. (you can look at todo examples e.g. here: https://github.com/yegor256/takes/search?q=%22%40todo%22+is%3Apr&type=Code&utf8=%E2%9C%93 or https://github.com/yegor256/netbout/search?utf8=%E2%9C%93&q=%22%40todo%22 or https://github.com/jcabi/jcabi-github/search?utf8=%E2%9C%93&q=%22%40todo%22&type=Code)

**Comment by krzyk** [comment]
@prahladyeri Also you have to add reference to the original issue in the issue description also (see http://at.teamed.io/qa.html)

**Comment by prahladyeri** [comment]
@krzyk thanks. I've edited the PR description and made a reference to the original issue. I've also added a puzzle in the javadoc of `PMDEmptyTest` class accordingly.

**Comment by krzyk** [comment]
@prahladyeri your `@todo` exceeds 80 column, please format it accordingly (remember that second line and all following, should have additional indentation of one space)

**Comment by prahladyeri** [comment]
@krzyk done. Sorry, I know about the 80 column rule, but since `qulice` didn't complain, I presumed that `javadoc` comments might be an exception.

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by davvd** [comment]
@carlosmiranda thanks, paid, **30 mins** to your account, payment ID is `74033650`, task took 168 hours and 14 mins to complete. review comments (c=15) added as a bonus. added +30 to your rating, now it is equal to [+2743](http://www.netbout.com/b/35023?open=rating)

**Comment by davvd** [comment]
@rultor deploy pls

**Pull request merged**